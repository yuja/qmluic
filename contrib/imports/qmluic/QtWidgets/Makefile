JQ = jq
QMAKE = qmake

QT_INSTALL_BINS = $(shell $(QMAKE) -query QT_INSTALL_BINS)
QT_INSTALL_LIBS = $(shell $(QMAKE) -query QT_INSTALL_LIBS)
QT_VERSION = $(shell $(QMAKE) -query QT_VERSION)

QMLTYPEREGISTRAR = $(QT_INSTALL_BINS)/qmltyperegistrar
QT_METATYPES = $(QT_INSTALL_LIBS)/metatypes

major_version = $(word 1,$(subst ., ,$(QT_VERSION)))
qtx = qt$(major_version)

.PHONY: all
all: plugins.qmltypes

plugins.qmltypes: metatypes.json $(QT_METATYPES)/$(qtx)core_metatypes.json
	$(QMLTYPEREGISTRAR) \
		--import-name qmluic.QtWidgets \
		--major-version $(major_version) \
		--foreign-types $(QT_METATYPES)/$(qtx)core_metatypes.json \
		--generate-qmltypes $@ $< > /dev/null

metatypes.json: $(QT_METATYPES)/$(qtx)widgets_metatypes.json
	$(JQ) 'map(.classes |= map({"classInfos": [{"name": "QML.Element", "value": "auto"}]} + .))' $< > $@

.PHONY: clean
clean:
	$(RM) plugins.qmltypes metatypes.json
