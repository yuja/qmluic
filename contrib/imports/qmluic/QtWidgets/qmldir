module qmluic.QtWidgets
typeinfo plugins.qmltypes
depends QtQml
