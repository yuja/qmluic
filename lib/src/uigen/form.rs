use super::object::UiObject;
use super::xmlutil;
use super::{FileNameRules, XmlResult, XmlWriter};
use crate::typemap::{Class, TypeSpace};
use quick_xml::events::{BytesStart, Event};
use std::io;

/// Top-level object wrapper to be serialized to UI XML.
#[derive(Clone, Debug)]
pub struct UiForm {
    pub class: Option<String>,
    pub root_object: UiObject,
    pub custom_widgets: Vec<CustomWidget>,
}

impl UiForm {
    /// Serializes this to UI XML.
    pub fn serialize_to_xml<W>(&self, writer: &mut XmlWriter<W>) -> XmlResult<()>
    where
        W: io::Write,
    {
        let tag = BytesStart::borrowed_name(b"ui").with_attributes([("version", "4.0")]);
        writer.write_event(Event::Start(tag.to_borrowed()))?;
        if let Some(name) = &self.class {
            xmlutil::write_tagged_str(writer, "class", name)?;
        }
        self.root_object.serialize_to_xml(writer)?;
        if !self.custom_widgets.is_empty() {
            let tag = BytesStart::borrowed_name(b"customwidgets");
            writer.write_event(Event::Start(tag.to_borrowed()))?;
            for w in &self.custom_widgets {
                w.serialize_to_xml(writer)?;
            }
            writer.write_event(Event::End(tag.to_end()))?;
        }
        writer.write_event(Event::End(tag.to_end()))?;
        writer.write(b"\n")
    }
}

/// User type referenced from the [`UiForm`].
#[derive(Clone, Debug)]
pub struct CustomWidget {
    pub class: String,
    pub extends: String,
    pub header: String,
}

impl CustomWidget {
    pub(super) fn from_class(cls: &Class, file_name_rules: &FileNameRules) -> Option<Self> {
        // If super class doesn't exist, diagnostic message would have been emitted while
        // building the object representation. So returns silently.
        let super_cls = cls.public_super_classes().next()?;
        Some(CustomWidget {
            class: cls.qualified_cxx_name().into(),
            extends: super_cls.qualified_cxx_name().into(),
            header: file_name_rules.type_name_to_cxx_header_name(cls.name()),
        })
    }

    /// Serializes this to UI XML.
    pub fn serialize_to_xml<W>(&self, writer: &mut XmlWriter<W>) -> XmlResult<()>
    where
        W: io::Write,
    {
        let tag = BytesStart::borrowed_name(b"customwidget");
        writer.write_event(Event::Start(tag.to_borrowed()))?;
        xmlutil::write_tagged_str(writer, "class", &self.class)?;
        xmlutil::write_tagged_str(writer, "extends", &self.extends)?;
        xmlutil::write_tagged_str(writer, "header", &self.header)?;
        writer.write_event(Event::End(tag.to_end()))
    }
}
