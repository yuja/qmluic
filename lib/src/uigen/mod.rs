//! Qt user interface XML (.ui) generator.

use crate::diagnostic::{Diagnostic, Diagnostics};
use crate::qmlast::{UiImportSource, UiProgram};
use crate::qmldir;
use crate::qmldoc::UiDocument;
use crate::typemap::{Class, Module, ModuleData, ModuleId, Type, TypeMap, TypeSpace};
use thiserror::Error;

mod expr;
mod gadget;
mod layout;
mod object;
mod property;
mod xmlutil;

pub use self::expr::*; // re-export
pub use self::gadget::*; // re-export
pub use self::layout::*; // re-export
pub use self::object::*; // re-export

pub type XmlError = quick_xml::Error;
pub type XmlResult<T> = quick_xml::Result<T>;
pub type XmlWriter<W> = quick_xml::Writer<W>;

/// Builds `UiForm` from the given `doc`.
pub fn build(
    base_ctx: &BuildContext,
    doc: &UiDocument,
    diagnostics: &mut Diagnostics,
) -> Option<UiForm> {
    let program = diagnostics.consume_err(UiProgram::from_node(doc.root_node(), doc.source()))?;
    // TODO: do not build unnamed module space if directory module available
    let module_data = make_doc_module_data(doc, &program, base_ctx.type_map, diagnostics);
    let type_space = get_doc_type(base_ctx.type_map, doc)
        .unwrap_or_else(|| Type::Module(Module::new(&module_data, base_ctx.type_map)));
    let ctx = BuildDocContext::new(doc, type_space, base_ctx);
    let (obj, cls) =
        object::resolve_object_definition(&ctx, program.root_object_node(), diagnostics)?;
    let root_object =
        UiObject::from_object_definition(&ctx, &cls, &obj, ContainerKind::Any, diagnostics)?;
    Some(UiForm {
        class: doc.type_name().map(|s| s.to_owned()),
        root_object,
    })
}

fn get_doc_type<'a>(type_map: &'a TypeMap, doc: &UiDocument) -> Option<Type<'a>> {
    if let Some(doc_path) = doc.path() {
        let base_dir =
            qmldir::normalize_path(doc_path.parent().expect("doc path should point to file"));
        type_map
            .get_module(ModuleId::Directory(base_dir.into()))
            .and_then(|m| {
                m.get_type(
                    doc.type_name()
                        .expect("doc loaded from file should have type"),
                )
            })
    } else {
        None
    }
}

fn make_doc_module_data(
    doc: &UiDocument,
    program: &UiProgram,
    type_map: &TypeMap,
    diagnostics: &mut Diagnostics,
) -> ModuleData {
    let mut module_data = ModuleData::with_builtins();
    for imp in program.imports() {
        if imp.alias().is_some() {
            diagnostics.push(Diagnostic::error(
                imp.node().byte_range(),
                "aliased import is not supported",
            ));
            continue;
        }
        // TODO: warn that version field is ignored
        // TODO: anchor diagnostic message onto imp.source() node
        let id = match imp.source() {
            UiImportSource::Identifier(x) => {
                ModuleId::Named(x.to_string(doc.source()).into_owned().into())
            }
            UiImportSource::String(x) => {
                if let Some(p) = doc.path().and_then(|p| p.parent()) {
                    ModuleId::Directory(qmldir::normalize_path(p.join(&x)).into())
                } else {
                    diagnostics.push(Diagnostic::error(
                        imp.node().byte_range(),
                        "cannot resolve directory path against inline QML document",
                    ));
                    continue;
                }
            }
        };
        if type_map.contains_module(&id) {
            module_data.import_module(id);
        } else {
            diagnostics.push(Diagnostic::error(
                imp.node().byte_range(),
                "module not found",
            ));
        }
    }
    module_data
}

/// Resources passed around the UI object constructors.
#[derive(Clone, Debug)]
pub struct BuildContext<'a> {
    type_map: &'a TypeMap,
    action_class: Class<'a>,
    form_layout_class: Class<'a>,
    grid_layout_class: Class<'a>,
    hbox_layout_class: Class<'a>,
    layout_class: Class<'a>,
    layout_attached_class: Class<'a>,
    push_button_class: Class<'a>,
    spacer_item_class: Class<'a>,
    tab_widget_class: Class<'a>,
    tab_widget_attached_class: Class<'a>,
    vbox_layout_class: Class<'a>,
    widget_class: Class<'a>,
}

#[derive(Clone, Debug)]
struct BuildDocContext<'a, 's> {
    source: &'s str,
    action_class: Class<'a>,
    form_layout_class: Class<'a>,
    grid_layout_class: Class<'a>,
    hbox_layout_class: Class<'a>,
    layout_class: Class<'a>,
    layout_attached_class: Class<'a>,
    push_button_class: Class<'a>,
    spacer_item_class: Class<'a>,
    tab_widget_class: Class<'a>,
    tab_widget_attached_class: Class<'a>,
    vbox_layout_class: Class<'a>,
    widget_class: Class<'a>,
    type_space: Type<'a>,
}

impl<'a> BuildContext<'a> {
    /// Sets up context with the given `type_map`.
    pub fn prepare(type_map: &'a TypeMap) -> Result<Self, BuildContextError> {
        const MODULE_NAME: &str = "qmluic.QtWidgets";
        let module = type_map
            .get_module(&ModuleId::Named(MODULE_NAME.into()))
            .ok_or(BuildContextError::ModuleNotFound(MODULE_NAME))?;
        let get_class = |name| {
            if let Some(Type::Class(cls)) = module.get_type(name) {
                Ok(cls)
            } else {
                Err(BuildContextError::ClassNotFound(name))
            }
        };
        Ok(BuildContext {
            type_map,
            action_class: get_class("QAction")?,
            form_layout_class: get_class("QFormLayout")?,
            grid_layout_class: get_class("QGridLayout")?,
            hbox_layout_class: get_class("QHBoxLayout")?,
            layout_class: get_class("QLayout")?,
            layout_attached_class: get_class("QLayoutAttached")?,
            push_button_class: get_class("QPushButton")?,
            spacer_item_class: get_class("QSpacerItem")?,
            tab_widget_class: get_class("QTabWidget")?,
            tab_widget_attached_class: get_class("QTabWidgetAttached")?,
            vbox_layout_class: get_class("QVBoxLayout")?,
            widget_class: get_class("QWidget")?,
        })
    }
}

impl<'a, 's> BuildDocContext<'a, 's> {
    fn new(doc: &'s UiDocument, type_space: Type<'a>, base_ctx: &BuildContext<'a>) -> Self {
        BuildDocContext {
            source: doc.source(),
            action_class: base_ctx.action_class.clone(),
            form_layout_class: base_ctx.form_layout_class.clone(),
            grid_layout_class: base_ctx.grid_layout_class.clone(),
            hbox_layout_class: base_ctx.hbox_layout_class.clone(),
            layout_class: base_ctx.layout_class.clone(),
            layout_attached_class: base_ctx.layout_attached_class.clone(),
            push_button_class: base_ctx.push_button_class.clone(),
            spacer_item_class: base_ctx.spacer_item_class.clone(),
            tab_widget_class: base_ctx.tab_widget_class.clone(),
            tab_widget_attached_class: base_ctx.tab_widget_attached_class.clone(),
            vbox_layout_class: base_ctx.vbox_layout_class.clone(),
            widget_class: base_ctx.widget_class.clone(),
            type_space,
        }
    }
}

/// Error occurred while setting up [`BuildContext`].
#[derive(Clone, Debug, Error)]
pub enum BuildContextError {
    #[error("required class not found: {0}")]
    ClassNotFound(&'static str),
    #[error("required module not found: {0}")]
    ModuleNotFound(&'static str),
}
