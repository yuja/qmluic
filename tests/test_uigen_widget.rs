use qmluic::uigen::DynamicBindingHandling;

pub mod common;

#[test]
fn test_root_must_be_widget() {
    insta::assert_snapshot!(common::translate_str(r###"
    import qmluic.QtWidgets
    QVBoxLayout {}
    "###).unwrap_err(), @r###"
    error: class 'QVBoxLayout' is not a QWidget
      ┌─ <unknown>:2:1
      │
    2 │ QVBoxLayout {}
      │ ^^^^^^^^^^^^^^
    "###);
}

#[test]
fn test_duplicated_id() {
    insta::assert_snapshot!(common::translate_str(r###"
    import qmluic.QtWidgets
    QWidget {
        QWidget { id: foo }
        QWidget { id: foo }
    }
    "###).unwrap_err(), @r###"
    error: duplicated object id: foo
      ┌─ <unknown>:4:19
      │
    3 │     QWidget { id: foo }
      │                   --- id is first defined here
    4 │     QWidget { id: foo }
      │                   ^^^ duplicated id is defined here
    "###);
}

#[test]
fn test_reserved_word_workaround() {
    insta::assert_snapshot!(common::translate_str(r###"
    import qmluic.QtWidgets
    QPushButton { default_: true }
    "###).unwrap(), @r###"
    <ui version="4.0">
     <class>MyType</class>
     <widget class="QPushButton" name="pushButton">
      <property name="default">
       <bool>true</bool>
      </property>
     </widget>
    </ui>
    "###);
}

#[test]
fn test_width_is_readonly() {
    insta::assert_snapshot!(common::translate_str(r###"
    import qmluic.QtWidgets
    QWidget { width: 100 }
    "###).unwrap_err(), @r###"
    error: not a writable property
      ┌─ <unknown>:2:11
      │
    2 │ QWidget { width: 100 }
      │           ^^^^^^^^^^
    "###);
}

#[test]
fn test_dynamic_width_is_readonly() {
    let doc = common::parse_doc(
        r###"
        import qmluic.QtWidgets
        QWidget {
            width: spinBox.value
            QSpinBox { id: spinBox }
        }
        "###,
    );
    insta::assert_snapshot!(
        common::translate_doc(&doc, DynamicBindingHandling::Generate).unwrap_err(),
        @r###"
    error: not a writable property
      ┌─ <unknown>:3:5
      │
    3 │     width: spinBox.value
      │     ^^^^^^^^^^^^^^^^^^^^
    "###);
}

#[test]
fn test_unobservable_property() {
    let doc = common::parse_doc(
        r###"
        import qmluic.QtWidgets
        QWidget {
             id: root
             QLabel { text: "width: %1".arg(root.width) }
        }
        "###,
    );
    insta::assert_snapshot!(
        common::translate_doc(&doc, DynamicBindingHandling::Generate).unwrap_err(),
        @r###"
    error: unobservable property: width
      ┌─ <unknown>:4:37
      │
    4 │      QLabel { text: "width: %1".arg(root.width) }
      │                                     ^^^^
    "###);
}

#[test]
fn test_incompatible_return_type() {
    let doc = common::parse_doc(
        r###"
        import qmluic.QtWidgets
        QWidget {
            windowTitle: {
                switch (edit.text) {
                case "hello":
                    return "world";
                }
            }
            QLineEdit { id: edit }
        }
        "###,
    );
    insta::assert_snapshot!(
        common::translate_doc(&doc, DynamicBindingHandling::Generate).unwrap_err(), @r###"
    error: cannot deduce return type from 'QString' and 'void'
      ┌─ <unknown>:8:6
      │
    6 │             return "world";
      │                    ------- type: QString
    7 │         }
    8 │     }
      │      ^ type: void
    "###);
}

#[test]
fn test_assign_double() {
    insta::assert_snapshot!(common::translate_str(r###"
    import qmluic.QtWidgets
    QDoubleSpinBox { value: 1.0 }
    "###).unwrap(), @r###"
    <ui version="4.0">
     <class>MyType</class>
     <widget class="QDoubleSpinBox" name="doubleSpinBox">
      <property name="value">
       <number>1</number>
      </property>
     </widget>
    </ui>
    "###);
}

#[test]
fn test_assign_qreal() {
    insta::assert_snapshot!(common::translate_str(r###"
    import qmluic.QtWidgets
    QDial { notchTarget: 3.7 }
    "###).unwrap(), @r###"
    <ui version="4.0">
     <class>MyType</class>
     <widget class="QDial" name="dial">
      <property name="notchTarget">
       <number>3.7</number>
      </property>
     </widget>
    </ui>
    "###);
}

#[test]
fn test_assign_float_to_int() {
    insta::assert_snapshot!(common::translate_str(r###"
    import qmluic.QtWidgets
    QSpinBox { value: 1.0 }
    "###).unwrap_err(), @r###"
    error: expression type mismatch (expected: int, actual: double)
      ┌─ <unknown>:2:19
      │
    2 │ QSpinBox { value: 1.0 }
      │                   ^^^
    "###);
}

#[test]
fn test_object_property_binding() {
    let doc = common::parse_doc(
        r###"
        import qmluic.QtWidgets
        QWidget {
             QCheckBox { id: source }
             QWidget { visible: source.checked }
        }
        "###,
    );
    let (ui_xml, ui_support_h) =
        common::translate_doc(&doc, DynamicBindingHandling::Generate).unwrap();
    insta::assert_snapshot!(ui_xml);
    insta::assert_snapshot!(ui_support_h);
}

#[test]
fn test_object_property_binding_to_root() {
    let doc = common::parse_doc(
        r###"
        import qmluic.QtWidgets
        QDialog {
             windowTitle: edit.text
             QLineEdit { id: edit }
        }
        "###,
    );
    let (ui_xml, ui_support_h) =
        common::translate_doc(&doc, DynamicBindingHandling::Generate).unwrap();
    insta::assert_snapshot!(ui_xml);
    insta::assert_snapshot!(ui_support_h);
}

#[test]
fn test_object_property_binding_unsupported() {
    insta::assert_snapshot!(common::translate_str(r###"
    import qmluic.QtWidgets
    QWidget {
         QCheckBox { id: source }
         QWidget { visible: source.checked }
    }
    "###).unwrap_err(), @r###"
    error: unsupported dynamic binding
      ┌─ <unknown>:4:25
      │
    4 │      QWidget { visible: source.checked }
      │                         ^^^^^^^^^^^^^^
    "###);
}

#[test]
fn test_implicit_this_property_binding() {
    let doc = common::parse_doc(
        r###"
        import qmluic.QtWidgets
        QCheckBox { checked: windowTitle === "" }
        "###,
    );
    let (_, ui_support_h) = common::translate_doc(&doc, DynamicBindingHandling::Generate).unwrap();
    insta::assert_snapshot!(ui_support_h);
}

#[test]
fn test_explicit_this_property_binding() {
    let doc = common::parse_doc(
        r###"
        import qmluic.QtWidgets
        QCheckBox { checked: this.windowTitle === "" }
        "###,
    );
    let (_, ui_support_h) = common::translate_doc(&doc, DynamicBindingHandling::Generate).unwrap();
    insta::assert_snapshot!(ui_support_h);
}

#[test]
fn test_indirect_property_binding() {
    let doc = common::parse_doc(
        r###"
        import qmluic.QtWidgets
        QWidget {
            QCheckBox { id: source }
            QWidget {
                visible: {
                    let w = source;
                    w.checked
                }
            }
        }
        "###,
    );
    let (_, ui_support_h) = common::translate_doc(&doc, DynamicBindingHandling::Generate).unwrap();
    insta::assert_snapshot!(ui_support_h);
}

#[test]
fn test_dynamic_property_source() {
    let doc = common::parse_doc(
        r###"
        import qmluic.QtWidgets
        QWidget {
            windowTitle: (source.checked ? edit1 : edit2).text
            QCheckBox { id: source }
            QLineEdit { id: edit1 }
            QLineEdit { id: edit2 }
        }
        "###,
    );
    let (_, ui_support_h) = common::translate_doc(&doc, DynamicBindingHandling::Generate).unwrap();
    insta::assert_snapshot!(ui_support_h);
}

#[test]
fn test_dynamic_binding_type_mismatch() {
    let doc = common::parse_doc(
        r###"
        import qmluic.QtWidgets
        QWidget {
             windowTitle: source.checked
             QCheckBox { id: source }
        }
        "###,
    );
    insta::assert_snapshot!(
        common::translate_doc(&doc, DynamicBindingHandling::Generate).unwrap_err(), @r###"
    error: expression type mismatch (expected: QString, actual: bool)
      ┌─ <unknown>:3:19
      │
    3 │      windowTitle: source.checked
      │                   ^^^^^^^^^^^^^^
    "###);
}

#[test]
fn test_omit_dynamic_binding() {
    let doc = common::parse_doc(
        r###"
        import qmluic.QtWidgets
        QWidget {
             QCheckBox { id: source }
             QWidget { visible: source.checked }
        }
        "###,
    );
    let (ui_xml, _) = common::translate_doc(&doc, DynamicBindingHandling::Omit).unwrap();
    insta::assert_snapshot!(ui_xml, @r###"
    <ui version="4.0">
     <class>MyType</class>
     <widget class="QWidget" name="widget1">
      <widget class="QCheckBox" name="source">
      </widget>
      <widget class="QWidget" name="widget">
      </widget>
     </widget>
    </ui>
    "###);
}

#[test]
fn test_deduplicate_dynamic_binding_senders() {
    let doc = common::parse_doc(
        r###"
        import qmluic.QtWidgets
        QDialog {
             windowTitle: combo.currentIndex === 0 ? "-" : combo.currentText
             QComboBox { id: combo }
        }
        "###,
    );
    let (_, ui_support_h) = common::translate_doc(&doc, DynamicBindingHandling::Generate).unwrap();
    insta::assert_snapshot!(ui_support_h);
}

#[test]
fn test_deduplicate_dynamic_binding_connections() {
    let doc = common::parse_doc(
        r###"
        import qmluic.QtWidgets
        QDialog {
             windowTitle: edit.text + edit.text
             QLineEdit { id: edit }
        }
        "###,
    );
    let (_, ui_support_h) = common::translate_doc(&doc, DynamicBindingHandling::Generate).unwrap();
    insta::assert_snapshot!(ui_support_h);
}

#[test]
fn test_deduplicate_dynamic_binding_sender_receiver() {
    let doc = common::parse_doc(
        r###"
        import qmluic.QtWidgets
        QLineEdit {
            id: edit
            enabled: edit.text === ""
        }
        "###,
    );
    let (_, ui_support_h) = common::translate_doc(&doc, DynamicBindingHandling::Generate).unwrap();
    insta::assert_snapshot!(ui_support_h);
}

#[test]
fn test_dynamic_binding_with_static_cast() {
    let doc = common::parse_doc(
        r###"
        import qmluic.QtWidgets
        QWidget {
            QDoubleSpinBox { id: edit }
            QComboBox { currentIndex: edit.value as int }
        }
        "###,
    );
    let (_, ui_support_h) = common::translate_doc(&doc, DynamicBindingHandling::Generate).unwrap();
    insta::assert_snapshot!(ui_support_h);
}

#[test]
fn test_integer_arithmetic_type_mismatch() {
    insta::assert_snapshot!(common::translate_str(r###"
    import qmluic.QtWidgets
    QWidget {
         windowTitle: "sum: %1".arg(edit1.value + edit2.value)
         QSpinBox { id: edit1 }
         QDoubleSpinBox { id: edit2 }
    }
    "###).unwrap_err(), @r###"
    error: operation '+' on incompatible types: int and double
      ┌─ <unknown>:3:33
      │
    3 │      windowTitle: "sum: %1".arg(edit1.value + edit2.value)
      │                                 -----------   ----------- type: double
      │                                 │              
      │                                 type: int
      │
      = use (expr as double) or (expr as int) for numeric cast
    "###);
}

#[test]
fn test_object_comparison_type_mismatch() {
    insta::assert_snapshot!(common::translate_str(r###"
    import qmluic.QtWidgets
    QWidget {
         visible: combo !== edit
         QComboBox { id: combo }
         QLineEdit { id: edit }
    }
    "###).unwrap_err(), @r###"
    error: operation '!=' on incompatible types: QComboBox* and QLineEdit*
      ┌─ <unknown>:3:15
      │
    3 │      visible: combo !== edit
      │               -----     ---- type: QLineEdit*
      │               │          
      │               type: QComboBox*
      │
      = use (expr as QWidget) to upcast to base class
    "###);
}

#[test]
fn test_ternary_expression_type_mismatch() {
    insta::assert_snapshot!(common::translate_str(r###"
    import qmluic.QtWidgets
    QWidget {
         windowTitle: source.checked ? 1 : "whatever"
         QCheckBox { id: source }
    }
    "###).unwrap_err(), @r###"
    error: operation 'ternary' on incompatible types: integer and QString
      ┌─ <unknown>:3:36
      │
    3 │      windowTitle: source.checked ? 1 : "whatever"
      │                                    -   ---------- type: QString
      │                                    │    
      │                                    type: integer
    "###);
}

#[test]
fn test_qstring_arg() {
    let doc = common::parse_doc(
        r###"
        import qmluic.QtWidgets
        QWidget {
            QLabel { text: "dynamic %1".arg(combo.currentIndex) }
            QLabel { text: "static %1".arg(1) }
            QComboBox { id: combo }
        }
        "###,
    );
    let (_, ui_support_h) = common::translate_doc(&doc, DynamicBindingHandling::Generate).unwrap();
    insta::assert_snapshot!(ui_support_h);
}

#[test]
fn test_nullable_property_sources() {
    let doc = common::parse_doc(
        r###"
        import qmluic.QtWidgets
        QLabel {
            text: {
                let w = check.checked ? edit : null;
                w !== null ? w.text : ""
            }
            QCheckBox { id: check }
            QLineEdit { id: edit }
        }
        "###,
    );
    let (_, ui_support_h) = common::translate_doc(&doc, DynamicBindingHandling::Generate).unwrap();
    insta::assert_snapshot!(ui_support_h);
}

#[test]
fn test_unique_binding_method_name() {
    let doc = common::parse_doc(
        r###"
        import qmluic.QtWidgets
        QWidget {
            id: main
            windowTitle: edit1.text  // "main" + "WindowTitle"
            QGroupBox {
                id: mainWindow
                title: edit2.text  // "mainWindow" + "Title"
            }
            QLineEdit { id: edit1 }
            QLineEdit { id: edit2 }
        }
        "###,
    );
    let (_, ui_support_h) = common::translate_doc(&doc, DynamicBindingHandling::Generate).unwrap();
    insta::assert_snapshot!(ui_support_h);
}

#[test]
fn test_dynamic_tab_widget_attached() {
    let doc = common::parse_doc(
        r###"
        import qmluic.QtWidgets
        QTabWidget {
            QWidget {
                id: tab
                QTabWidget.title: tab.windowTitle
            }
        }
        "###,
    );
    insta::assert_snapshot!(
        common::translate_doc(&doc, DynamicBindingHandling::Generate).unwrap_err(), @r###"
    error: unused or unsupported dynamic binding to attached property
      ┌─ <unknown>:5:9
      │
    5 │         QTabWidget.title: tab.windowTitle
      │         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    "###);
}

#[test]
fn test_static_actions_list() {
    insta::assert_snapshot!(common::translate_str(r###"
    import qmluic.QtWidgets
    QMenu {
        actions: [act, menu.menuAction(), sep]
        QAction { id: act }
        QMenu { id: menu }
        QAction { id: sep; separator: true }
    }
    "###).unwrap(), @r###"
    <ui version="4.0">
     <class>MyType</class>
     <widget class="QMenu" name="menu1">
      <addaction name="act"/>
      <addaction name="menu"/>
      <addaction name="separator"/>
      <action name="act">
      </action>
      <widget class="QMenu" name="menu">
      </widget>
     </widget>
    </ui>
    "###);
}

#[test]
fn test_actions_list_incompatible_type() {
    insta::assert_snapshot!(common::translate_str(r###"
    import qmluic.QtWidgets
    QMenu {
        actions: [
            act1,
            act2,
            menu,
        ]
        QAction { id: act1 }
        QAction { id: act2 }
        QMenu { id: menu }
    }
    "###).unwrap_err(), @r###"
    error: incompatible array element types at index 2: QAction* and QMenu*
      ┌─ <unknown>:5:9
      │
    5 │         act2,
      │         ---- type: QAction*
    6 │         menu,
      │         ---- type: QMenu*
      │
      = call .menuAction() to obtain QAction* associated with menu
    "###);
}

#[test]
fn test_action_separator_false() {
    insta::assert_snapshot!(common::translate_str(r###"
    import qmluic.QtWidgets
    QMenu {
        QAction { separator: false }
    }
    "###).unwrap(), @r###"
    <ui version="4.0">
     <class>MyType</class>
     <widget class="QMenu" name="menu">
      <addaction name="action"/>
      <action name="action">
      </action>
     </widget>
    </ui>
    "###);
}

#[test]
fn test_action_seaparator_dynamic() {
    let doc = common::parse_doc(
        r###"
        import qmluic.QtWidgets
        QWidget {
            QCheckBox { id: check }
            QAction { separator: check.checked }
        }
        "###,
    );
    let (ui_xml, ui_support_h) =
        common::translate_doc(&doc, DynamicBindingHandling::Generate).unwrap();
    insta::assert_snapshot!(ui_xml, @r###"
    <ui version="4.0">
     <class>MyType</class>
     <widget class="QWidget" name="widget">
      <addaction name="action"/>
      <widget class="QCheckBox" name="check">
      </widget>
      <action name="action">
      </action>
     </widget>
    </ui>
    "###);
    insta::assert_snapshot!(ui_support_h);
}

#[test]
fn test_action_seaparator_with_other_properties() {
    let doc = common::parse_doc(
        r###"
        import qmluic.QtWidgets
        QWidget {
            QAction { separator: true; text: "whatever" }
        }
        "###,
    );
    let (ui_xml, ui_support_h) =
        common::translate_doc(&doc, DynamicBindingHandling::Generate).unwrap();
    insta::assert_snapshot!(ui_xml, @r###"
    <ui version="4.0">
     <class>MyType</class>
     <widget class="QWidget" name="widget">
      <addaction name="action"/>
      <action name="action">
       <property name="text">
        <string notr="true">whatever</string>
       </property>
      </action>
     </widget>
    </ui>
    "###);
    insta::assert_snapshot!(ui_support_h);
}

#[test]
fn test_action_seaparator_with_callback() {
    let doc = common::parse_doc(
        r###"
        import qmluic.QtWidgets
        QWidget {
            QAction { separator: true; onHovered: {} }
        }
        "###,
    );
    let (ui_xml, ui_support_h) =
        common::translate_doc(&doc, DynamicBindingHandling::Generate).unwrap();
    insta::assert_snapshot!(ui_xml, @r###"
    <ui version="4.0">
     <class>MyType</class>
     <widget class="QWidget" name="widget">
      <addaction name="action"/>
      <action name="action">
      </action>
     </widget>
    </ui>
    "###);
    insta::assert_snapshot!(ui_support_h);
}
