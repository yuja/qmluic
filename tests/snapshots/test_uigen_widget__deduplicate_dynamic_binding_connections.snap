---
source: tests/test_uigen_widget.rs
expression: ui_support_h
---
#pragma once
#include "ui_mytype.h"

namespace UiSupport {
class MyType
{
public:
    MyType(QDialog *root, Ui::MyType *ui): root_(root), ui_(ui) {}

    void setup()
    {
        this->setup_dialog_windowTitle();
        this->update_dialog_windowTitle();
    }

private:
    void setup_dialog_windowTitle()
    {
        auto *const edit = this->ui_->edit;

        QObject::connect(edit, &QLineEdit::textChanged, root_, [this]() { this->update_dialog_windowTitle(); });
    }

    void update_dialog_windowTitle()
    {
        this->root_->setWindowTitle(this->eval_dialog_windowTitle());
    }

    QString eval_dialog_windowTitle() const
    {
        QString a0;
        QString a1;
        QString a2;
    b0:
        a0 = this->ui_->edit->text();
        a1 = this->ui_->edit->text();
        a2 = a0 + a1;
        return a2;
    }

    QDialog *const root_;
    Ui::MyType *const ui_;

};
} // namespace UiSupport

